### PR TITLE
Map GW LoA result back

### PR DESCRIPTION
**Pull request by pablothedude**
Map the LoA back from GW to EB value.

https://www.pivotaltracker.com/story/show/166729672

### CODE REVIEW

File: src/OpenConext/EngineBlockBundle/Stepup/StepupGatewayLoaMapping.php [hunk 1]
```diff
@@ -57,6 +57,21 @@ public function transformToGatewayLoa($input)
...
+
+    /**
+     * @param $input
+     * @return string
+     */
+    public function transformToEbLoa($input)
+    {
+        $loa = array_search($input, $this->mapping);
+        if ($loa === false) {
+            throw new RuntimeException('Unable to determine manage LOA for gateway LOA');
```
Comment (MKodde) [inline src/OpenConext/EngineBlockBundle/Stepup/StepupGatewayLoaMapping.php]: This error message needs a revision! Suggestion:

`Unable to map the received stepup LoA back to the configured EngineBlock LoA`

File: library/EngineBlock/Corto/Module/Service/StepupAssertionConsumer.php [hunk 2]
```diff
@@ -72,6 +72,11 @@ public function serve($serviceName, Request $httpRequest)
         try {
             $receivedResponse = $this->_server->getBindingsModule()->receiveResponse($serviceEntityId, $expectedDestination);
             $receivedRequest = $this->_server->getReceivedRequestFromResponse($receivedResponse);
+
+            // Update the AuthnContextClassRef to the loa returned
+            $processStep = $this->_processingStateHelper->getStepByRequestId($receivedRequest->getId(), ProcessingStateHelperInterface::STEP_STEPUP);
```
[the remark this answers is not in the document]
Comment (pablothedude) [reply library/EngineBlock/Corto/Module/Service/StepupAssertionConsumer.php]: Good suggestion

File: library/EngineBlock/Corto/Module/Service/StepupAssertionConsumer.php [hunk 3]
```diff
@@ -72,6 +72,11 @@ public function serve($serviceName, Request $httpRequest)
...
+            $processStep->getResponse()->getAssertion()->setAuthnContextClassRef($this->_stepupGatewayCallOutHelper->getEbLoa($receivedResponse->getAssertion()->getAuthnContextClassRef()));
+            $this->_processingStateHelper->updateStepResponseByRequestId($receivedRequest->getId(), ProcessingStateHelperInterface::STEP_STEPUP, $processStep->getResponse());
```
[the remark this answers is not in the document]
Comment (pablothedude) [reply library/EngineBlock/Corto/Module/Service/StepupAssertionConsumer.php]: Good suggestion, I have moved it to a separate function

File: src/OpenConext/EngineBlockBundle/Stepup/StepupGatewayLoaMapping.php [hunk 4]
```diff
@@ -57,6 +57,21 @@ public function transformToGatewayLoa($input)
         return $this->mapping[$input];
     }
 
+
+    /**
+     * @param $input
+     * @return string
+     */
+    public function transformToEbLoa($input)
```
[the remark this answers is not in the document]
Comment (pablothedude) [reply src/OpenConext/EngineBlockBundle/Stepup/StepupGatewayLoaMapping.php]: That was the reason why the WIP-label was there, will add it

File: src/OpenConext/EngineBlockBundle/Stepup/StepupGatewayLoaMapping.php (hunk 1 above)
Comment (pablothedude) [reply src/OpenConext/EngineBlockBundle/Stepup/StepupGatewayLoaMapping.php]: Good suggestion, done

File: src/OpenConext/EngineBlockBundle/Stepup/StepupGatewayLoaMapping.php [hunk 5]
```diff
@@ -51,12 +51,27 @@ public function __construct(array $loaMapping, $gatewayLoa1)
     public function transformToGatewayLoa($input)
     {
         if (!array_key_exists($input, $this->mapping)) {
-            throw new RuntimeException('Unable to determine gateway LOA for manage LOA');
+            throw new RuntimeException('Unable to map the EngineBlock LoA to a configured stepup LoA');
```
Comment (MKodde) [inline src/OpenConext/EngineBlockBundle/Stepup/StepupGatewayLoaMapping.php]: Very nice addition, but the exception message could be more specific, you could mention that the input loa is not found in the configured loa mapping.

File: tests/unit/OpenConext/EngineBlockBundle/Stepup/StepupGatewayLoaMappingTest.php [hunk 6]
```diff
@@ -89,4 +90,39 @@ public function the_stepup_loa_mapping_loa1_should_be_a_string_or_throw_an_excep
...
     }
+
+    /**
+     * @test
+     * @group Stepup
+     */
+    public function the_stepup_loa_mapping_object_should_successful_map_back()
+    {
+        $stepupLoaMapping = new StepupGatewayLoaMapping([
+            'manageLoa2' => 'gatewayLoa2',
```
Comment (MKodde) [inline tests/unit/OpenConext/EngineBlockBundle/Stepup/StepupGatewayLoaMappingTest.php]: manageLoa? It's no big deal, just curiosity!

**Pull request merged**